sonic_platform_base/fan_base.py

```diff
@@ -0,0 +1,100 @@
+#
+# fan_base.py
+#
+# Abstract base class for implementing a platform-specific class with which
+# to interact with a fan module in SONiC
+#
+
+try:
+    import abc
+    from . import device_base
+except ImportError as e:
+    raise ImportError(str(e) + " - required module not found")
+
+
+class FanBase(device_base.DeviceBase):
+    """
+    Abstract base class for interfacing with a fan module
+    """
+    __metaclass__ = abc.ABCMeta
+    
+    # Possible fan directions
+    FAN_DIRECTION_INTAKE = "intake"
+    FAN_DIRECTION_EXHAUST = "exhaust"
+
+    # Possible fan status LED colors
+    STATUS_LED_COLOR_GREEN = "green"
+    STATUS_LED_COLOR_RED = "red"
+    STATUS_LED_COLOR_OFF = "off"
+
+    @abc.abstractmethod
+    def get_direction(self):
+        """
+        Retrieves the direction of fan
+
+        Returns:
+            A string, either FAN_DIRECTION_INTAKE or FAN_DIRECTION_EXHAUST
+            depending on fan direction
+        """
+        return None
+
+    @abc.abstractmethod
+    def get_speed(self):
+        """
+        Retrieves the speed of fan as a percentage of full speed
+
+        Returns:
+            An integer, the percentage of full fan speed, in the range 0 (off)
+                 to 100 (full speed)
+        """
+        return 0
+
+    @abc.abstractmethod
+    def get_target_speed(self):
+        """
+        Retrieves the target (expected) speed of the fan
+
+        Returns:
+            An integer, the percentage of full fan speed, in the range 0 (off)
+                 to 100 (full speed)
+        """
+        return 0
+
+    @abc.abstractmethod
+    def get_speed_tolerance(self):
+        """
+        Retrieves the speed tolerance of the fan
+
+        Returns:
+            An integer, the percentage of variance from target speed which is
+                 considered tolerable
+        """
+        return 0
+
+    @abc.abstractmethod
+    def set_speed(self, speed):
+        """
+        Sets the fan speed
+
+        Args:
+            speed: An integer, the percentage of full fan speed to set fan to,
```

Comment:
I see everywhere that the fan speed is an int from 0 to 100.
Some device dealing with fans can have finer grained control (like 0 to 255) which means loosing a bit of precision.
Was this taken into consideration and integer deemed better than floats?

Reply:
Would you prefer a 0.0 ... 1.0 float scale? This should allow for finer-grained control if necessary, while maintaining a more universal scale.

Reply:
I don't mind either way, integer percentages 0 to 100 sounds good to me.
I just wanted to bring this topic up as the [sysfs interface](https://www.kernel.org/doc/Documentation/hwmon/sysfs-interface) defines a pwm value to be ranging from 0 to 255.

Reply:
If all platforms used sysfs, the 0-255 scale would be fine. However, we'd like a universal scale that will work with other interfaces (IPMI, etc.).